File: libraries/AdaptiveExpressions/BuiltinFunctions/And.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using AdaptiveExpressions.Memory;

namespace AdaptiveExpressions.BuiltinFunctions
{
    /// <summary>
    /// Check whether all expressions are true. Return true if all expressions are true,
    /// or return false if at least one expression is false.
    /// </summary>
    public class And : ExpressionEvaluator
    {
        public And()
            : base(ExpressionType.And, Evaluator, ReturnType.Boolean, FunctionUtils.ValidateAtLeastOne)
        {
        }

        private static (object value, string error) Evaluator(Expression expression, IMemory state, Options options)
        {
            object result = true;
            string error = null;
            foreach (var child in expression.Children)
            {
                (result, error) = child.TryEvaluate(state, new Options(options) { NullSubstitution = null });
                if (error == null)
                {
                    if (FunctionUtils.IsLogicTrue(result))
                    {
                        result = true;
                    }
                    else
                    {
                        result = false;
                        break;
                    }
                }
                else
                {
                    // We interpret any error as false and swallow the error
                    result = false;
                    error = null;
                    break;
                }
            }

            return (result, error);
        }
    }
}

File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/BuiltinFunctions/And.cpp
#include "And.h"
#include "../Code/ExpressionType.h"
#include "../Code/FunctionUtils.h"

#include <limits>
#include <any>
#include <cstdlib>

AdaptiveExpressions_BuiltinFunctions::And::And() :
    ExpressionEvaluator(ExpressionType::And, (ReturnType)((int)ReturnType::String | (int)ReturnType::Number))
{
}

ValueErrorTuple AdaptiveExpressions_BuiltinFunctions::And::TryEvaluate(Expression* expression, void* state, Options* options)
{
    std::any result = true;
    std::string error;

    // Create a childEvaluateOptions using the locale from the passed in parameter but with NullSubstitution unset
    Options childEvaluateOptions;
    childEvaluateOptions.locale = options ? options->locale : "";

    for (int i = 0; i < expression->getChildrenCount(); ++i)
    {
        Expression* child = expression->getChildAt(i);

        ValueErrorTuple childResult = child->TryEvaluate(state, &childEvaluateOptions);
        if (childResult.second.empty())
        {
            if (FunctionUtils::isLogicTrue(childResult.first))
            {
                result = true;
            }
            else
            {
                result = false;
                break;
            }
        }
        else
        {
            // We interpret any error as false and swallow the error
            result = false;
            error = nullptr;
            break;
        }
    }

    return ValueErrorTuple(result, error);
}

void AdaptiveExpressions_BuiltinFunctions::And::ValidateExpression(Expression* expression)
{
    FunctionUtils::ValidateArityAndAnyType(expression, 1, INT_MAX);
}